core/vcpu.c

```diff
@@ -2536,50 +2548,16 @@ static void handle_cpuid_virtual(struct vcpu_t *vcpu, uint32_t a, uint32_t c)
     uint32_t hw_family;
     uint32_t hw_model;
     uint8_t physical_address_size;
+    uint32_t cpuid_1_features_ecx, cpuid_1_features_edx,
+             cpuid_8000_0001_features_ecx, cpuid_8000_0001_features_edx;
 
-    static uint32_t cpuid_1_features_edx =
-            FEATURE(PAT)        |
-            FEATURE(FPU)        |
-            FEATURE(VME)        |
-            FEATURE(DE)         |
-            FEATURE(TSC)        |
-            FEATURE(MSR)        |
-            FEATURE(PAE)        |
-            FEATURE(MCE)        |
-            FEATURE(CX8)        |
-            FEATURE(APIC)       |
-            FEATURE(SEP)        |
-            FEATURE(MTRR)       |
-            FEATURE(PGE)        |
-            FEATURE(MCA)        |
-            FEATURE(CMOV)       |
-            FEATURE(CLFSH)      |
-            FEATURE(MMX)        |
-            FEATURE(FXSR)       |
-            FEATURE(SSE)        |
-            FEATURE(SSE2)       |
-            FEATURE(SS)         |
-            FEATURE(PSE)        |
-            FEATURE(HTT);
-
-    static uint32_t cpuid_1_features_ecx =
-            FEATURE(SSE3)       |
-            FEATURE(SSSE3)      |
-            FEATURE(SSE41)      |
-            FEATURE(SSE42)      |
-            FEATURE(CMPXCHG16B) |
-            FEATURE(MOVBE)      |
-            FEATURE(AESNI)      |
-            FEATURE(PCLMULQDQ)  |
-            FEATURE(POPCNT);
-
-    static uint32_t cpuid_8000_0001_features_edx =
-            FEATURE(NX)         |
-            FEATURE(SYSCALL)    |
-            FEATURE(RDTSCP)     |
-            FEATURE(EM64T);
-
-    static uint32_t cpuid_8000_0001_features_ecx = 0;
+    // To fully support CPUID instructions (opcode = 0F A2) by software, it is
+    // recommended to add opcode_table_0FA2[] in core/emulate.c to emulate
+    // (Refer to Intel SDM Vol. 2A 3.2 CPUID).
+    cpuid_get_guest_features(vcpu->guest_cpuid, &cpuid_1_features_ecx,
+                             &cpuid_1_features_edx,
+                             &cpuid_8000_0001_features_ecx,
+                             &cpuid_8000_0001_features_edx);
```

Comment:
SET_CPUID interface allows to set all output registers for any cpuid codes. But the only used data is ecx and edx for code 1 and 80000001. Why?

Reply:
Good question. Other function code can be added one by one in future. We did not open the setting channels for all codes as HAXM still has not supported for all features in current implementation.

Reply:
I think https://github.com/intel/haxm/pull/271 should be called set_cpuid and this feature set_cpuid_feature_flags. How Android Studio will use this feature? Who adds ioctl calls there?